lisa/tools/free.py

```diff
@@ -4,20 +4,78 @@
+# Swap:        4194304           0     4194304
+
 
 class Free(Tool):
     @property
     def command(self) -> str:
         return "free"
 
-    def get_swap_size(self) -> int:
-        # Return total swap size
+    #     Usage:
+    #  free [options]
+
+    # Options:
+    #  -b, --bytes         show output in bytes
+    #      --kilo          show output in kilobytes
+    #      --mega          show output in megabytes
+    #      --giga          show output in gigabytes
+    #      --tera          show output in terabytes
+    #      --peta          show output in petabytes
+    #  -k, --kibi          show output in kibibytes
+    #  -m, --mebi          show output in mebibytes
+    #  -g, --gibi          show output in gibibytes
+    #      --tebi          show output in tebibytes
+    #      --pebi          show output in pebibytes
+    #  -h, --human         show human-readable output
+    #      --si            use powers of 1000 not 1024
+    #  -l, --lohi          show detailed low and high memory statistics
+    #  -t, --total         show total for RAM + swap
+    #  -s N, --seconds N   repeat printing every N seconds
+    #  -c N, --count N     repeat printing N times, then exit
+    #  -w, --wide          wide output
+
+    # Starter data sizes, use Kibi/Mebi/Gibi
+    _byte_size_flags = {"KiB": "-k", "MiB": "-m", "GiB": "-g"}
+
+    def _get_header_index(self, header_name: str, header_line: str) -> int:
+        # get index of header based on first line, add one to account
+        # for first label field in the rest of the lines
+        return 1 + header_line.split().index(header_name)
+
+    def _get_field(
+        self, field_name: str, header_name: str, byte_size: str = "MiB"
+    ) -> int:
+
         # Example output:
         #         total used free
         # Swap:   0     0    0
-        out = self.run("-m", force_run=True).stdout
+
+        # fetch the flag for the data format we want
+        byte_size_flag = self._byte_size_flags[byte_size]
+
+        # get the data
+        out = self.run(byte_size_flag, force_run=True).stdout
+        lines = out.splitlines()
+
+        # get offset for data when we split the line
+        header_index = self._get_header_index(header_name, lines[0])
+
         for line in out.splitlines():
-            if line.startswith("Swap:"):
-                return int(line.split()[1])
+            if line.startswith(f"{field_name}:"):
+                return int(line.split()[header_index])
+
+        raise LisaException(f"Failed to get info for field {field_name}")
+
+    def get_swap_size(self) -> int:
+        # Return total swap size in Mebibytes
+        return self._get_field("Swap", "total", "MiB")
+
+    def get_free_memory_mb(self) -> int:
+        return self._get_field("Mem", "free", "MiB")
 
-        raise LisaException("Failed to get swap size")
+    def get_free_memory_gb(self) -> int:
+        return self._get_field("Mem", "free", "GiB")
```

Comment:
Responding to previous review question from @squirrelsc, I usually try to just wrap the tool's behavior.

Reply:
Actually digging in, maybe it makes more sense to just use the default and then shift? We'll lose granularity below KiB but the behavior and style of flags has changed over time, while KibiBytes have always been the default.

Reply:
yeah I'm switching it, good note thank you 🙂

Reply:
It's ok to keep KB level. We're not wrapping the tool for general purpose use, and it's for test cases of LISA.

Reply:
It was reviewing the git history that did it for me, wanting to avoid flag issues across versions